### PR TITLE
Competition

### PR DESCRIPTION
**Pull request by yyz561**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

mmdetection implementation of CenterNet2.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/models/losses/focal_loss.py [hunk 1]
```diff
@@ -179,3 +179,67 @@ def forward(self,
         else:
             raise NotImplementedError
         return loss_cls
+
+
+@LOSSES.register_module()
+class BinaryFocalLoss(nn.Module):
+
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/focal_loss.py:186]: Need add DocSring at here,

File: mmdet/models/losses/focal_loss.py [hunk 2]
```diff
@@ -179,3 +179,67 @@ def forward(self,
...
+    def __init__(self,
+                 alpha: float = 0.25,
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/focal_loss.py:188]: using docstring, and do not need to add float here.

File: mmdet/models/losses/focal_loss.py [hunk 3]
```diff
@@ -179,3 +179,67 @@ def forward(self,
...
+        self.pos_weight = pos_weight
+        self.neg_weight = neg_weight
+        self.sigmoid_clamp = sigmoid_clamp
+        self.ignore_high_fp = ignore_high_fp
+
+    def forward(self, inputs, targets, pos_inds, avg_factor=None):
+
+        """
+        Args:
+            inputs:  (sum_l N*Hl*Wl,)
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/focal_loss.py:211]: add type and other details

File: mmdet/models/losses/focal_loss.py [hunk 4]
```diff
@@ -179,3 +179,67 @@ def forward(self,
...
+        super(BinaryFocalLoss, self).__init__()
+        self.alpha = alpha
+        self.beta = beta
+        self.gamma = gamma
+        self.pos_weight = pos_weight
+        self.neg_weight = neg_weight
+        self.sigmoid_clamp = sigmoid_clamp
+        self.ignore_high_fp = ignore_high_fp
+
+    def forward(self, inputs, targets, pos_inds, avg_factor=None):
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/focal_loss.py:207]: add reduction_override=None to make loss applicable, more details can refer to #5515

File: mmdet/models/necks/bifpn.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+from mmcv.runner import BaseModule
+from ..builder import NECKS
+
+
+class SingleBiFPN(BaseModule):
+    """
+    Single layer of BiFPN
+    """
+    def __init__(self,
+                 in_channels_list,
```
Comment (BIGWangYuDong) [inline mmdet/models/necks/bifpn.py:14]: in_channels

File: mmdet/models/necks/bifpn.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn import ConvModule
+from mmcv.runner import BaseModule
+from ..builder import NECKS
+
+
+class SingleBiFPN(BaseModule):
+    """
```
Comment (BIGWangYuDong) [inline mmdet/models/necks/bifpn.py:10]: add args docstring

**Comment by BIGWangYuDong** [comment]
need add test unit

**Comment by jshilong** [comment]
Thanks for your great works, Have you benchmarked this pr and get similar performance with the original implementation?

File: mmdet/models/necks/bifpn.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+                 act_cfg=None,
+                 init_cfg=dict(
+                     type='Normal',
+                     layer='Conv2d',
+                     std=0.01,)):
+        super(BiFPN, self).__init__(init_cfg)
+        assert isinstance(in_channels, list)
+        assert num_outs == 5 or num_outs == 3, NotImplementedError
+        self.num_bifpn = num_bifpn
+        self.__extra_levels = nn.ModuleList()
```
Comment (jshilong) [inline mmdet/models/necks/bifpn.py:173]: The prefix  `—` seems unnecessary. In BiFPN, I believe following the name in FPN is a better choice.

File: mmdet/models/necks/bifpn.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+                    nn.Parameter(
+                        torch.ones(len(inputs_offsets),
+                                   dtype=torch.float32), requires_grad=True))
+
+            # generate convolutions after combination
+            self.add_module(f"outputs_f{feat_level}_{inputs_offsets_str}",
+                            ConvModule(out_channels,
+                                       out_channels,
+                                       kernel_size=3,
+                                       padding=1,
```
Comment (jshilong) [inline mmdet/models/necks/bifpn.py:71]: Does this module need custom `init_weights`?

File: mmdet/models/losses/focal_loss.py [hunk 9]
```diff
@@ -179,3 +179,67 @@ def forward(self,
         else:
             raise NotImplementedError
         return loss_cls
+
+
+@LOSSES.register_module()
```
Comment (jshilong) [inline mmdet/models/losses/focal_loss.py:184]: Is there any difference with 
https://github.com/open-mmlab/mmdetection/blob/87eda061c466238cf2b746ec749777144499ba20/mmdet/models/losses/focal_loss.py#L106
If no, please use the version in mmdet

File: mmdet/models/detectors/centernet2.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+
+@DETECTORS.register_module()
+class CenterNet2(TwoStageDetector):
+    """Implementation of `Probabilistic two-stage detection
+        <https://arxiv.org/abs/2103.07461>`_.
+
+    Slightly different from original models:
+    the difference is:
+    Stage 1: Original CenterNet2 when calculating heatmap loss,
```
Comment (jshilong) [inline mmdet/models/detectors/centernet2.py:12]: Will this affect the performance

File: mmdet/models/necks/bifpn.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+class SingleBiFPN(BaseModule):
+    """
+    Single layer of BiFPN
+    """
+    def __init__(self,
+                 in_channels_list,
+                 out_channels,
+                 conv_cfg=None,
+                 norm_cfg=None,
+                 act_cfg=None):
```
Comment (AronLin) [inline mmdet/models/necks/bifpn.py:18]: Can we add arguments like FPN to conveniently control this function?
like `num_outs`, `start_level`, `end_level`, `in_channels`, `out_channels`

File: mmdet/models/losses/focal_loss.py (hunk 3 above)
Comment (yyz561) [reply mmdet/models/losses/focal_loss.py:211]: Hey, thanks for the comment. Currently there still has a ton of work to do with this implementation. However, I am in the competition and the due is around the corner, so I made the PR before further polishing. Most importantly, I haven't had a decent result yet (it is still training on a single GPU). So I think I will address these comments once I have some meaningful results. Thanks again.

File: mmdet/models/necks/bifpn.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+                    output_feats.append(feats[-1 - i])
+                    break
+            else:
+                raise ValueError()
+        return output_feats
+
+
+@NECKS.register_module()
+class BiFPN(BaseModule):
+    """ The neck used in 'CenterNet2 <http://arxiv.org/abs/2103.07461>'_
```
[the remark this answers is not in the document]
Comment (yyz561) [reply mmdet/models/necks/bifpn.py:134]: Hey, thanks for the comment. Currently there still has a ton of work to do with this implementation. However, I am in the competition and the due is around the corner, so I made the PR before further polishing. Most importantly, I haven't had a decent result yet (it is still training on a single GPU). So I think I will address these comments once I have some meaningful results. Thanks again.

File: mmdet/models/necks/bifpn.py (hunk 11 above)
Comment (yyz561) [reply mmdet/models/necks/bifpn.py:18]: Hey, thanks for the comment. Currently there still has a ton of work to do with this implementation. However, I am in the competition and the due is around the corner, so I made the PR before further polishing. Most importantly, I haven't had a decent result yet (it is still training on a single GPU). So I think I will address these comments once I have some meaningful results. Thanks again.

File: mmdet/models/detectors/centernet2.py (hunk 10 above)
Comment (LMerCy) [reply mmdet/models/detectors/centernet2.py:12]: @jshilong Centernet2 should be support faster-rcnn-retinanet, so instead of create a new centernet2head, create a new two-stage detector to give labels to one-stage head maybe a better practice?

Comment (yyz561) [reply mmdet/models/detectors/centernet2.py:12]: yeah, will look carefully at FCOS head and RetinaNet head closely to see whether we can build the head on top of them.

Comment (LMerCy) [reply mmdet/models/detectors/centernet2.py:12]: @yyz561  Actually you can refactor you centernet head loss function to be same with fcos head, then create a new a new two stage, to give them same input. May  i write the new two stage in your branch?

Comment (yyz561) [reply mmdet/models/detectors/centernet2.py:12]: Sure, welcome to participate in this work!

**Comment by yyz561** [comment]
> Thanks for your great works, Have you benchmarked this pr and get similar performance with the original implementation?

Hello, I have the sults fi-fi-finally (sigh...). A quick result report please refer to my teamates' result report "Extra" part. Link: https://github.com/huimlight/mmdet-centernetv2/blob/main/README.md

File: mmdet/models/losses/focal_loss.py (hunk 9 above)
Comment (yyz561) [reply mmdet/models/losses/focal_loss.py:184]: I believe there are some small differences. This version introduced a relatively limited number if positive sample points, and only used these points to calculate positve loss.
